CRM/Activity/Selector/Activity.php

```diff
@@ -417,11 +417,11 @@ public function &getRows($action, $offset, $rowCount, $sort, $output = NULL, $ca
         $row['engagement_level'] = CRM_Utils_Array::value($engagementLevel, $engagementLevels, $engagementLevel);
       }
 
-      $actionLinks = $this->actionLinks(CRM_Utils_Array::value('activity_type_id', $row),
-        CRM_Utils_Array::value('source_record_id', $row),
+      $actionLinks = $this->actionLinks($row['activity_type_id'],
+        $row['source_record_id'] ?? NULL,
```

Comment:
There should be two `??` expressions here?

Reply:
Actually I think there could be zero.  The `$row` variable comes from `CRM_Activity_BAO_Activity::getActivities()` which is hard-coded to always return both `'activity_type_id'` and `'source_record_id'` keys, so they will never be undefined.